pkg/k8s/watchers/endpointsynchronizer.go

```diff
@@ -282,5 +288,70 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
 					return nil
 				}
 			},
+			StopFunc: func(ctx context.Context) error {
+				podName := e.GetK8sPodName()
+				if podName == "" {
+					scopedLog.Debug("Skipping CiliumEndpoint deletion because it has no k8s pod name")
+					return nil
+				}
+				namespace := e.GetK8sNamespace()
+				if namespace == "" {
+					scopedLog.Debug("Skipping CiliumEndpoint deletion because it has no k8s namespace")
+					return nil
+				}
+				if err := ciliumClient.CiliumEndpoints(namespace).Delete(ctx, podName, meta_v1.DeleteOptions{}); err != nil {
+					if !k8serrors.IsNotFound(err) {
+						scopedLog.WithError(err).Warning("Unable to delete CEP")
+					}
+				}
+				return nil
+			},
 		})
 }
+
+// DeleteK8sCiliumEndpointSync replaces the endpoint controller to remove the
+// CEP from Kubernetes once the endpoint is stopped / removed from the
+// Cilium agent.
+func (epSync *EndpointSynchronizer) DeleteK8sCiliumEndpointSync(e *endpoint.Endpoint) {
+	controllerName := controllerNameOf(e.ID)
+
+	scopedLog := e.Logger(subsysEndpointSync).WithField("controller", controllerName)
+
+	if !k8s.IsEnabled() {
+		scopedLog.Debug("Not starting controller because k8s is disabled")
+		return
+	}
+
+	ciliumClient := k8s.CiliumClient().CiliumV2()
+
+	// The health endpoint doesn't really exist in k8s and updates to it caused
+	// arbitrary errors. Disable the controller for these endpoints.
+	if isHealthEP := e.HasLabels(pkgLabels.LabelHealth); isHealthEP {
+		scopedLog.Debug("Not starting unnecessary CEP controller for cilium-health endpoint")
+		return
+	}
+
+	// NOTE: The controller functions do NOT hold the endpoint locks
+	e.UpdateController(controllerName,
+		controller.ControllerParams{
+			StopFunc: func(ctx context.Context) error {
+				podName := e.GetK8sPodName()
+				if podName == "" {
+					scopedLog.Debug("Skipping CiliumEndpoint deletion because it has no k8s pod name")
+					return nil
+				}
+				namespace := e.GetK8sNamespace()
+				if namespace == "" {
+					scopedLog.Debug("Skipping CiliumEndpoint deletion because it has no k8s namespace")
+					return nil
+				}
+				if err := ciliumClient.CiliumEndpoints(namespace).Delete(ctx, podName, meta_v1.DeleteOptions{}); err != nil {
+					if !k8serrors.IsNotFound(err) {
+						scopedLog.WithError(err).Warning("Unable to delete CEP")
```

Comment:
I guess this is best effort, ie if there's another error like lack of permissions then we will just give up rather than returning an error to the controller to ensure it gets retried?